src/libOpenImageIO/imageoutput.cpp

```diff
@@ -177,8 +177,14 @@ ImageOutput::write_tiles(int xbegin, int xend, int ybegin, int yend, int zbegin,
                     ok &= write_tile(x, y, z, format, tilestart, xstride,
                                      ystride, zstride);
                 } else {
-                    if (!buf.get())
-                        buf.reset(new char[pixelsize * m_spec.tile_pixels()]);
+                    if (!buf.get()) {
+                        const size_t sz = pixelsize * m_spec.tile_pixels();
+                        char* mem       = new char[sz];
+                        // Not all pixels will be initialized, so we set them to zero here.
+                        // This will avoid generation of NaN, FPEs and valgrind errors.
+                        memset(mem, 0, sz);
+                        buf.reset(mem);
+                    }
```

Comment:
I'm confused about where the access of the uninitialized memory is. In copy_image? In write_tile?

Also, I wonder what you think about the following slightly alternate idiom:

                    if (!buf.get()) {
                        const size_t sz = pixelsize * m_spec.tile_pixels();
                        buf.reset(new char[sz]);
                        // Not all pixels will be initialized, so we set them to zero here.
                        // This will avoid generation of NaN, FPEs and valgrind errors.
                        memset(buf.get(), 0, sz);
                    }

I think this is a smaller change from the previous code's approach and doesn't need to introduce an additional raw pointer variable?

Reply:
The callstack of the use of uninitialized data is in the bug report of issue #4462

Note that valgrind gives you two callstacks if you issue the `--track-origins=yes` option:

First it shows you where the access of garbage data happened.

Then it shows you where the data in question was allocated.

The invalid data is used in `parallel_convert_from_float()` when it converts **all** 64x64 FP pixels, including the ones that were never assigned a value. The actual leaf-func that triggers a branch on invalid data (and thus is UB) is the rounding function `roundf()`.

I will replace the explicit allocation variable as you suggested, but I find them equally valid.

Reply:
Oh, and I see what you mean about the confusion... so the UB was triggered not in the main thread, but in a thread from the thread pool. So it is not traced all the way back to `main()`

(That's why I was asking about a mode that uses no threading at all, and does all the work on the main thread. Sadly, there is no such mode: Even with `--threads 1` the work will be performed on a separate thread. I think it is better form to have a switch to disable all threading, even if it is only for easier debugger sessions. But fixing that is probably non trivial and outside the scope of this issue.)

Reply:
It surprises me that --threads 1 doesn't run all in the main thread. I wonder if there is someplace that is failing to honor that number and using the full thread ppool.

Reply:
What about if in addition to --threads 1, you set env `OPENIMAGEIO_THREADS=1` ?

Reply:
Yeah, I consider this a bug.